paimon-flink/paimon-flink-common/src/main/java/org/apache/paimon/flink/lookup/DynamicPartitionLevelLoader.java

```diff
@@ -0,0 +1,195 @@
+                        .sorted(comparator.reversed())
+                        .collect(Collectors.toList());
+
+        if (maxPartitionLoadLevel == table.partitionKeys().size() - 1) {
+            // if maxPartitionLoadLevel is the max partition level, we only need to load the max
+            // partition
+            if (newPartitions.size() <= 1) {
+                return newPartitions;
+            } else {
+                return newPartitions.subList(0, 1);
+            }
+        }
+
+        for (int level = 0; level <= maxPartitionLoadLevel; level++) {
+            newPartitions = extractMaxPartitionsForFixedLevel(newPartitions, level);
+            if (LOG.isDebugEnabled()) {
+                LOG.debug(
+                        "DynamicPartitionLevelLoader(currentPartitionLoadLevel={},table={}) finds new partitions: {}.",
+                        level,
+                        table.name(),
+                        partitionsToString(newPartitions));
+            }
+        }
+        return newPartitions;
+    }
+
+    private int getMaxPartitionLoadLevel(
+            Map<String, String> partitionLoadConfig, List<String> partitionFields) {
+        Preconditions.checkArgument(partitionLoadConfig.size() <= partitionFields.size());
+
+        int maxLoadLevel = -1;
+        for (int i = 0; i < partitionFields.size(); i++) {
+            String config = partitionLoadConfig.getOrDefault(partitionFields.get(i), null);
+
+            if (i == 0) {
+                Preconditions.checkArgument(
+                        config != null, "the top level partition must set load config.");
+            }
+            if (config != null) {
+                Preconditions.checkArgument(
+                        config.equals(MAX_PT),
+                        "unsupported load config '{}' for partition field '{}', currently only support '{}'.",
+                        config,
+                        partitionFields.get(i),
+                        MAX_PT);
+                maxLoadLevel = Math.max(maxLoadLevel, i);
+
+            } else {
+                // if level-i partition don't set config, we won't consider the partition level
+                // which is greater than level-i
+                return maxLoadLevel;
+            }
+        }
+        return maxLoadLevel;
+    }
+
+    private List<InternalRow.FieldGetter> createPartitionFieldGetters() {
+        List<InternalRow.FieldGetter> fieldGetters = new ArrayList<>();
+
+        RowType partitionType = table.rowType().project(table.partitionKeys());
+
+        for (int i = 0; i < maxPartitionLoadLevel + 1; i++) {
+            fieldGetters.add(InternalRow.createFieldGetter(partitionType.getTypeAt(i), i));
+        }
+        return fieldGetters;
+    }
+
+    private List<BinaryRow> extractMaxPartitionsForFixedLevel(
+            List<BinaryRow> partitions, int level) {
+        AtomicInteger currentDistinct = new AtomicInteger(0);
```

Comment:
Why do we use AtomicInteger here? Seems that there is no parallel execution issue here.